daemon/volumes.go

```diff
@@ -275,60 +159,142 @@ func copyExistingContents(source, destination string) error {
 		if err != nil {
 			return err
 		}
 
-		if err := mount.Mount(source, destPath, "bind", "rbind,rw"); err != nil {
-			return fmt.Errorf("error while mounting volume %s: %v", source, err)
+		for _, m := range c.MountPoints {
+			cp := m
+			cp.RW = mode != "ro"
+
+			if len(m.Source) == 0 {
+				v, err := createVolume(m.Name, m.Driver)
+				if err != nil {
+					return err
+				}
+				cp.Volume = v
+			}
+
+			mountPoints[cp.Destination] = cp
 		}
 	}
 
-	for _, mnt := range container.specialMounts() {
-		destPath, err := container.GetResourcePath(mnt.Destination)
+	// 3. Read bind mounts
+	for _, b := range hostConfig.Binds {
+		// #10618
+		bind, err := parseBindMount(b, container.Config)
 		if err != nil {
 			return err
 		}
-		if err := mount.Mount(mnt.Source, destPath, "bind", "bind,rw"); err != nil {
-			return fmt.Errorf("error while mounting volume %s: %v", mnt.Source, err)
+
+		if binds[bind.Destination] {
+			return fmt.Errorf("Duplicate bind mount %s", bind.Destination)
 		}
+
+		if len(bind.Name) > 0 && len(bind.Driver) > 0 {
+			v, err := createVolume(bind.Name, bind.Driver)
+			if err != nil {
+				return err
+			}
+			bind.Volume = v
+		}
+
+		binds[bind.Destination] = true
+		mountPoints[bind.Destination] = bind
 	}
+
+	container.MountPoints = mountPoints
+
 	return nil
 }
 
-func (container *Container) unmountVolumes() {
-	for dest := range container.Volumes {
-		destPath, err := container.GetResourcePath(dest)
-		if err != nil {
-			logrus.Errorf("error while unmounting volumes %s: %v", destPath, err)
-			continue
-		}
-		if err := mount.ForceUnmount(destPath); err != nil {
-			logrus.Errorf("error while unmounting volumes %s: %v", destPath, err)
-			continue
+// verifyOldVolumesInfo ports volumes configured for the containers pre docker 1.7.
+// It reads the container configuration and creates valid mount points for the old volumes.
+func (daemon *Daemon) verifyOldVolumesInfo(container *Container) error {
+	jsonPath, err := container.jsonPath()
+	if err != nil {
+		return err
+	}
+	f, err := os.Open(jsonPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
 		}
+		return err
 	}
 
-	for _, mnt := range container.specialMounts() {
-		destPath, err := container.GetResourcePath(mnt.Destination)
-		if err != nil {
-			logrus.Errorf("error while unmounting volumes %s: %v", destPath, err)
-			continue
-		}
-		if err := mount.ForceUnmount(destPath); err != nil {
-			logrus.Errorf("error while unmounting volumes %s: %v", destPath, err)
+	type oldContVolCfg struct {
+		Volumes   map[string]string
+		VolumesRW map[string]bool
+	}
+
+	var vols oldContVolCfg
+	if err := json.NewDecoder(f).Decode(&vols); err != nil {
+		return err
+	}
+
+	for destination, hostPath := range vols.Volumes {
+		vfsPath := filepath.Join(daemon.root, "vfs", "dir")
+
+		if strings.HasPrefix(hostPath, vfsPath) {
+			id := filepath.Base(hostPath)
+
+			container.AddLocalMountPoint(id, destination, vols.VolumesRW[destination])
```

Comment:
thanks for caching that issue @cpuguy83! I've change the initialization to handle this problem.